[ARM][Performance]Improve ARM CPU depthwise convolution performance

FrozenGene opened the pull request:
To leverage existing spatial pack schedule and add tunable compute_at knob to re-implement ARM CPU's depthwise convolution. 

On my A53@2.0GHz ARM CPU (MTK6763), which can boost 1.6X performance compared with previous depthwise convolution in the Mobilenet V1 model (I have also checked the correctness of this schedule). 

The following is the [Tensorflow Mobilenet V1](http://download.tensorflow.org/models/mobilenet_v1_2018_02_22/mobilenet_v1_1.0_224.tgz) model auto tvm training GFLOPS log:
Currently:
[Task 2/20] Current/Best: 0.98/ 2.32 GFLOPS | Progress: (1427/2000) | 2679.82 s Done.
[Task 4/20] Current/Best: 0.56/ 1.15 GFLOPS | Progress: (1072/2000) | 2461.27 s Done.
[Task 6/20] Current/Best: 1.08/ 2.78 GFLOPS | Progress: (1084/2000) | 1987.91 s Done.
[Task 8/20] Current/Best: 0.39/ 1.19 GFLOPS | Progress: (1815/2000) | 2744.70 s Done.
[Task 10/20] Current/Best: 1.09/ 2.33 GFLOPS | Progress: (1222/2000) | 1866.02 s Done.
[Task 12/20] Current/Best: 0.42/ 0.90 GFLOPS | Progress: (1716/2000) | 2528.94 s Done.
[Task 14/20] Current/Best: 1.89/ 2.63 GFLOPS | Progress: (1284/2000) | 2288.55 s Done.
[Task 16/20] Current/Best: 0.47/ 0.96 GFLOPS | Progress: (1467/2000) | 2282.65 s Done.
[Task 18/20] Current/Best: 1.43/ 2.61 GFLOPS | Progress: (1007/2000) | 1525.76 s Done.

After this PR optimization:
[Task 2/20] Current/Best: 0.00/ 4.83 GFLOPS | Progress: (1682/2000) | 1470.40 s Done.
[Task 4/20] Current/Best: 1.35/ 3.17 GFLOPS | Progress: (1257/2000) | 1032.80 s Done.
[Task 6/20] Current/Best: 2.04/ 5.49 GFLOPS | Progress: (1904/2000) | 1623.10 s Done.
[Task 8/20] Current/Best: 0.75/ 3.15 GFLOPS | Progress: (1885/2000) | 1546.22 s Done.
[Task 10/20] Current/Best: 2.09/ 6.07 GFLOPS | Progress: (2000/2000) | 1640.41 s Done.
[Task 12/20] Current/Best: 2.99/ 3.80 GFLOPS | Progress: (1853/2000) | 1547.13 s Done.
[Task 14/20] Current/Best: 4.59/ 6.06 GFLOPS | Progress: (1355/2000) | 1091.93 s Done.
[Task 16/20] Current/Best: 1.96/ 4.01 GFLOPS | Progress: (2000/2000) | 1586.18 s Done.
[Task 18/20] Current/Best: 2.33/ 4.63 GFLOPS | Progress: (2000/2000) | 1599.89 s Done.

The depthwise convolution total execution time on single A53@2.0GHz time can be from `45.3839ms` to `28.1945ms`. 

One thing you must notice to use this schedule: You MUST make the XGBTunner constructor’s feature type argument be feature_type= 'knob'. i.e.  XGBTuner(tsk, loss_type='rank', feature_type='knob'). Otherwise your program maybe hang forever.

@merrymercy @tqchen Pls review it.

FrozenGene commented:
The test_topi_depthwise_conv2d.py test error is because I have modified the schedule, which doesn't have `tile_c` and so on in previous schedule. But TopHub(fall back will use) doesn't contain my new schedule cfg. Maybe I need @merrymercy 's help to handle this situation.

merrymercy commented:
I am afraid that we have to put this PR on hold for a while. Since the current structure of TOPI is not easy to do such an update. But this can be a good example for other people who are intrested in ARM CPU.

Currently,
* We have tuned records (tophub) / benchmark for a lot of devices.
* We are doing some refactor to TOPI ( #1970 #1993 )

We will first finish the refactor, and then think about how to make such kind of update more easilly.

FrozenGene commented:
`We have tuned records (tophub) / benchmark for a lot of devices.`
Yes, we can not require every committer has so many devices. If committer change the schedule template, everything will be changed. How do we avoid it? Do you have any idea? I think we can not let you do this thing every time. 

I come up one way is we can have x86's fall_back schedule: https://github.com/dmlc/tvm/blob/master/topi/python/topi/x86/conv2d_avx_common.py#L12

tqchen commented:
I think the general idea moving forward, once we have a good idea of schedule template, is always create a new "fork" of an existing template when we make improvements, and keep the versioned template stable for two releases

merrymercy commented:
My proposal is to bring version number to autotvm template. So we will have template keys like `direct#0`, `direct#1`, `direct#4`.
Currently, we always use `direct` as the default/fallback template key. Users also use it to create tuning tasks. We can still keep this. But the system will redirect `direct` to the latest version.
In the log, we will log template key with the version number, so we can dispatch them to correct templates when using `apply_history_best`.

Another critical problem of this template is that XGBTuner with default parameters doesn't support tuning `compute_at`. I think I have to solve this problem before we pick this template as default template.

yzhliu commented on topi/python/topi/arm_cpu/depthwise_conv2d.py:
```diff
@@ -5,15 +5,140 @@
 
 # register customized schedule for arm cpu.
-@autotvm.register_topi_schedule(schedule_depthwise_conv2d_nchw, ['arm_cpu', 'cpu'], 'direct')
+@autotvm.task.register_topi_schedule(schedule_depthwise_conv2d_nchw, 'arm_cpu', 'direct')
```
@FrozenGene Would you mind also test how this change impact Intel CPU, as I described in #2045 ?

FrozenGene replied:
I will test it on my Mac cpu. What you want to compare? Previous ARM CPU Depthwise convolution schedule on x86 or your PR #2045? In fact, I tested NCHW[x]c schedule on ARM before, which doesn't have performance advantages on Mobilenet V1 model and even slower than previous depthwise convolution schedule.

yzhliu replied:
I would like to clarify. I'm not saying we should use NCHW[x]c on ARM. What I wondering is,
1. You remove the register for "cpu" here, it will make x86 use the default schedule for `depthwise_conv2d_nchw`, and users will see performance decreases. So we need to either add the register back, or make another for x86.
2. Then the question is, shall we rely on this schedule for x86, or make another nchw one? I see great performance decreasing on x86 with this schedule, compared to the previous one, both are using fallback config - so I suggest that, if you could provide the tuned fallback schedule, we can test it on x86 to see what is the better path forward.

FrozenGene replied:
Ok, I understand your concern. But my concern is that previous depthwise convolution fallback will load reference log: 
```python
        if cfg.is_fallback:
            ref_log = autotvm.tophub.load_reference_log(
                'arm_cpu', 'rk3399', 'depthwise_conv2d_nchw', 'direct')
            cfg.fallback_with_reference_log(ref_log)
```
However, we haven't uploaded such tuned log for this PR's schedule(I think this is the main reason of performance decreasing on x86). 

I can not make sure that whether I upload the tuned log for this PR's schedule before merge it for your fallback testing. But from my side, I can test that: tune for arm_cpu  using my schedule on mobilenet v1 and store it on one log file, then x86 use this log file to be fallback.

merrymercy replied:
We can not compare templates with fallback configs. So we can tell nothing from @yzlius 's test on x86.

The old template have good fallback mechanism, while the new template does not have any fallback mechanism. So in this case, the new template just gives naive implementation with no tiling, vectorization, etc.

FrozenGene replied:
@merrymercy yes, I understand your meaning. @yzhliu have added `cpu` on this new schedule template but have poorer performance on x86 with fallback config. However, to verify it, I want to do is to add one tuned log for arm_cpu using new template,  then make it be the standard of x86 fallback like old template has.

FrozenGene replied:
> It is not easy to do so automatically. You can just paste your log in a comment. But It also does not make much sense to test arm_cpu log on x86.
> So the only reasonable way for @yzhliu to verify is to tune by himself.

I know it. However, previous schedule does like this. Maybe we should have one `x86` cpu fallback log to reference, not `arm_cpu` log to reference like old schedule.

FrozenGene replied:
@yzhliu could you tune it use this schedule? I quickly test it on my x86 cpu, the first depthwise convolution gflops is 23.75GFLOPS, while previous schedule is 14.44GFLOPS. I think I would agree @merrymercy’s point,  it does not make sense use arm_cpu log to be fallback test on x86 cpu.

yzhliu replied:
Sure, I'll tune it today and report the number. Thanks for clarifying, I got a more clear picture now.

yzhliu replied:
With tuned kernel on x86, I got 5.53ms (O2) and 5.68ms (O3) on c5.9xlarge for mobilenet, while the previous one is 2.2ms (O2) and 2.7ms (O3). (It's normal to have O2 better than O3 in this case since O3 will convert non-depthwise conv2d to NCHW[x]c, thus introduces layout transformer between depthwise-conv and non-depthwise-conv)

Here's my script, config and running log:

https://gist.github.com/yzhliu/fe28f0e9add3cbd70dda4c7cd30c1275
https://gist.github.com/yzhliu/c287798592f7568c6e624b33727576a5
https://gist.github.com/yzhliu/dd7e16fbe3599ba5dce14986eb699d60

Let me know if I did something wrong. The LLVM errors in the running log looks a little bit weird.

I'll tune again tonight, coz I just realized I made a mistake that I tuned using only 1-thread (but evaluated with 18). Not sure whether that would make difference on final performance.

yzhliu replied:
@FrozenGene I'll try your option. How did you calculate gflops for mobilenet? it can be implied from running time, right?

FrozenGene replied:
@yzhliu I mean the tunning log's GFLOPS. For example, I list the GFLOPS here:
```
Currently:
[Task 2/20] Current/Best: 0.98/ 2.32 GFLOPS | Progress: (1427/2000) | 2679.82 s Done.

After this PR optimization:
[Task 2/20] Current/Best: 0.00/ 4.83 GFLOPS | Progress: (1682/2000) | 1470.40 s Done
```
Then I can compare the best GFLOPS between old / new schedule. And the running time should also affect this.

For the LLVM error, I also met this. You can redirect log python3 tune.py > gflops.log, without 2>& 1.

FrozenGene replied:
@yzhliu This gist file is on my x86 machine new schedule v.s. previous schedule on mobilenet v1's depthwise tunning GFLOPS log: https://gist.github.com/FrozenGene/2142bbe9c4df1ca307328a53cdac7b05

At my x86 machine, this new PR's schedule is better than old schedule. You could tune again after setting my option and correct thread nums.

yzhliu replied:
got it. thanks for clarifying. will do.

yzhliu replied:
Now it's 1.87ms (O2) and 2.09ms (O3), pretty cool!
new log: https://gist.github.com/yzhliu/c5b4ad171c0d8b862fe30bdad09434b3

I think then we can add the "cpu" register back, and upload this log to tophub.

FrozenGene replied:
Good. it is better than old depthwise schedule on x86 too. However, for the fallback, I think it is also not make much sense using arm_cpu log. So do you have suggestion on it? @merrymercy @yzhliu

yzhliu replied:
what do you mean by "arm_cpu" log?

FrozenGene replied:
> what do you mean by "arm_cpu" log?

I mean this code:
```python
        if cfg.is_fallback:
            ref_log = autotvm.tophub.load_reference_log(
                'arm_cpu', 'rk3399', 'depthwise_conv2d_nchw', 'direct')  # This is arm_cpu log
            cfg.fallback_with_reference_log(ref_log)
```
However, we use this log to be fallback reference log before.

yzhliu replied:
I believe we should have new models for fallback log, per cpu and arm_cpu

yzhliu commented:
@FrozenGene Would you further act on this PR?

FrozenGene commented:
@yzhliu fine. I will work on it on this week. I will add x86 cpu back based on latest code. If have anything I should notice, please let me know. CC @merrymercy

yzhliu commented:
@FrozenGene That's great. Thanks!

FrozenGene commented:
@yzhliu I have updated based on the latest code. 

I have done three things.

- added x86 cpu back

- dilated depthwise convolution support

- fixed depthwise conv2d infer shape issue to support we will split the channel of kernel

Please review it.

CC @merrymercy , because I have modifed the schedule template and maybe need your help to update the TopHub. Currently our CI test will meet key error `tile_co` in the: `test_top_level2.test_grouped_conv2d_nchw` and `test_topi_depthwise_conv2d.test_depthwise_conv2d`

yzhliu commented:
@merrymercy Could you help with the tophub?

merrymercy commented:
Currently, we cannot bring this in because

1. We have old tuned logs for many CPUs that are used in published benchmarks
2. The default parameters of XGBoostTuner does not support tuning this space.

I will work on solving 2. But 1 is not on my plan.

So if we want to accept this, a possible method is to name this template by another name, e.g., "contrib_spatial_pack" rather than "direct", so other users can use it by specifying template_name as we do for winograd.
After I fix 2, I may bring the version system (https://github.com/dmlc/tvm/pull/2028#issuecomment-434552438) to TOPI template and can set this as the default template.

eqy commented:
Do we plan to open some kind of RFC for the versioning? Right now it seems that there are many versions already (TVM release#, tophub distribution#, and now template#).

yzhliu commented:
+1 for bringing it into contrib package first, as @merrymercy suggested. @FrozenGene could you act accordingly? Also let's open an issue for tracking the status - version, XGBoostTuner, and benchmark.

FrozenGene commented:
> You understand it correctly. I think we can put it in the same file.

Thanks for replying. I will begin to do it no later than next week.

FrozenGene commented:
I'm very sorry that I commit the merge code previously. I wish this doesn't interrupt you.

Currently, I open one new PR: https://github.com/dmlc/tvm/pull/2345 to continue this work and add this PR as reference in case people are interested in the background. 

Sorry again for my mistake.
